### PR TITLE
Add tests for `sycl_ext_oneapi_composite_device` extension

### PR DESCRIPTION
**Pull request by AlexeySachkov**
Test plan: KhronosGroup/SYCL-CTS#865

### CODE REVIEW

File: tests/extension/oneapi_composite_device/aspects.cpp [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+  SKIP(
+      "The sycl_ext_oneapi_composite device extension is not supported by an "
+      "imlementation");
+#else
+  auto platform = sycl_cts::util::get_cts_object::platform();
+  auto platform_composite_devices = platform.ext_oneapi_get_composite_devices();
+  auto all_composite_devices =
+      sycl::ext::oneapi::experimental::get_composite_devices();
+
+  INFO(
```
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/aspects.cpp:45]: Should we also check that the devices returned by `platform::get_composite_devices` are present in the result of the free function `get_composite_devices`?

File: tests/extension/oneapi_composite_device/descendent_device.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+#include <vector>
+
+namespace composite_device::tests {
+
+TEST_CASE("Test for impact on descendent device",
+          "[oneapi_composite_device]") {
+#ifndef SYCL_EXT_ONEAPI_COMPOSITE_DEVICE
+  SKIP(
+      "The sycl_ext_oneapi_composite device extension is not supported by an "
+      "imlementation");
```
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/descendent_device.cpp:36]:
```suggestion
      "implementation");
```

File: tests/extension/oneapi_composite_device/aspects.cpp [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+#endif
+}
+
+TEST_CASE("Test for ext_oneapi_is_component aspect",
+          "[oneapi_composite_device]") {
+#ifndef SYCL_EXT_ONEAPI_COMPOSITE_DEVICE
+  SKIP(
+      "The sycl_ext_oneapi_composite device extension is not supported by an "
+      "imlementation");
```
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/aspects.cpp:84]:
```suggestion
      "implementation");
```

File: tests/extension/oneapi_composite_device/aspects.cpp [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+#include <vector>
+
+namespace composite_device::tests {
+
+TEST_CASE("Test for aspects of composite devices",
+          "[oneapi_composite_device]") {
+#ifndef SYCL_EXT_ONEAPI_COMPOSITE_DEVICE
+  SKIP(
+      "The sycl_ext_oneapi_composite device extension is not supported by an "
+      "imlementation");
```
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/aspects.cpp:38]:
```suggestion
      "implementation");
```

File: tests/extension/oneapi_composite_device/device_info_descriptors.cpp [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+#include <vector>
+
+namespace composite_device::tests {
+
+TEST_CASE("Test for info::device::component_devices descriptor",
+          "[oneapi_composite_device]") {
+#ifndef SYCL_EXT_ONEAPI_COMPOSITE_DEVICE
+  SKIP(
+      "The sycl_ext_oneapi_composite device extension is not supported by an "
+      "imlementation");
```
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/device_info_descriptors.cpp:37]:
```suggestion
      "implementation");
```

File: tests/extension/oneapi_composite_device/device_info_descriptors.cpp [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+#endif
+}
+
+TEST_CASE("Test for info::device::composite_device descriptor",
+          "[oneapi_composite_device]") {
+#ifndef SYCL_EXT_ONEAPI_COMPOSITE_DEVICE
+  SKIP(
+      "The sycl_ext_oneapi_composite device extension is not supported by an "
+      "imlementation");
```
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/device_info_descriptors.cpp:90]:
```suggestion
      "implementation");
```

File: tests/extension/oneapi_composite_device/device_info_descriptors.cpp [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+            "Querying a composite device should result in exception if query "
+            "is performed for a non-component device");
+      }
+
+      REQUIRE(result.has(sycl::aspect::ext_oneapi_is_composite));
+    } catch (const sycl::exception &e) {
+      if (!device.has(sycl::aspect::ext_oneapi_is_component)) {
+        INFO(
+            "Checking that a non-component device don't have a composite "
+            "device associcated with it");
```
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/device_info_descriptors.cpp:123]:
```suggestion
            "device associated with it");
```

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+#include <vector>
+
+namespace composite_device::tests {
+
+TEST_CASE("Test for signatures of the new get_composite_device APIs",
+          "[oneapi_composite_device]") {
+#ifndef SYCL_EXT_ONEAPI_COMPOSITE_DEVICE
+  SKIP(
+      "The sycl_ext_oneapi_composite device extension is not supported by an "
+      "imlementation");
```
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:38]:
```suggestion
      "implementation");
```

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+#endif
+}
+
+TEST_CASE("Test for free function get_composite_device",
+          "[oneapi_composite_device]") {
+#ifndef SYCL_EXT_ONEAPI_COMPOSITE_DEVICE
+  SKIP(
+      "The sycl_ext_oneapi_composite device extension is not supported by an "
+      "imlementation");
```
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:60]:
```suggestion
      "implementation");
```

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+#else
+
+  std::vector<sycl::device> composite_devices;
+
+  // get_composite_device may not throw
```
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:65]:
```suggestion
  // get_composite_devices may not throw
```

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+    composite_devices =
+        sycl::ext::oneapi::experimental::get_composite_devices();
+  } catch (sycl::exception& e) {
+    FAIL("get_composite_devices threw an exception: " << e.what());
+  }
+
+  {
+    INFO("Subsequent calls to get_composite_devices()");
+    std::vector<sycl::device> composite_devices2;
+    // get_composite_device may not throw
```
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:76]:
```suggestion
    // get_composite_devices may not throw
```

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+
+#endif
+}
+
+TEST_CASE("Test for platform::ext_oneapi_get_composite_devices",
+          "[oneapi_composite_device]") {
+#ifndef SYCL_EXT_ONEAPI_COMPOSITE_DEVICE
+  SKIP(
+      "The sycl_ext_oneapi_composite device extension is not supported by an "
+      "imlementation");
```
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:95]:
```suggestion
      "implementation");
```

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+#else
+
+  auto platform = sycl_cts::util::get_cts_object::platform();
+  std::vector<sycl::device> composite_devices;
+
+  // get_composite_device may not throw
```
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:101]:
```suggestion
  // get_composite_devices may not throw
```

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+    composite_devices = platform.ext_oneapi_get_composite_devices();
+  } catch (sycl::exception& e) {
+    FAIL("platform::ext_oneapi_get_composite_devices threw an exception: "
+         << e.what());
+  }
+
+  {
+    INFO("Subsequent calls to platform::ext_oneapi_get_composite_devices()");
+    std::vector<sycl::device> composite_devices2;
+    // get_composite_device may not throw
```
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:112]:
```suggestion
    // get_composite_devices may not throw
```

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+#endif
+}
+
+TEST_CASE(
+    "Test to check that composite devices are not returned as root devices",
+    "[oneapi_composite_device]") {
+#ifndef SYCL_EXT_ONEAPI_COMPOSITE_DEVICE
+  SKIP(
+      "The sycl_ext_oneapi_composite device extension is not supported by an "
+      "imlementation");
```
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:132]:
```suggestion
      "implementation");
```

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+#else
+
+  auto platform = sycl_cts::util::get_cts_object::platform();
+
+  std::vector<sycl::device> root_devices = platform.get_devices();
+  for (auto device : sycl::device::get_devices()) {
+    root_devices.push_back(device);
+  }
```
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:140]: Why do we need this? I thought `sycl::device::get_devices()` should already return the devices returned by `platform.get_devices()`, am I wrong?

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+
+  std::vector<sycl::device> composite_devices =
+      platform.ext_oneapi_get_composite_devices();
+  for (auto device : sycl::ext::oneapi::experimental::get_composite_devices()) {
+    composite_devices.push_back(device);
+  }
```
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:146]: Similar to the above one: the free function should include all the composite devices returned by `platform.ext_oneapi_get_composite_devices`, why do we need them both?

File: tests/extension/oneapi_composite_device/more_complex_test_cases.cpp [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+#include <vector>
+
+namespace composite_device::tests {
+
+TEST_CASE("Basic test for a composite device",
+          "[oneapi_composite_device]") {
+#ifndef SYCL_EXT_ONEAPI_COMPOSITE_DEVICE
+  SKIP(
+      "The sycl_ext_oneapi_composite device extension is not supported by an "
+      "imlementation");
```
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/more_complex_test_cases.cpp:36]:
```suggestion
      "implementation");
```

File: tests/extension/oneapi_composite_device/more_complex_test_cases.cpp [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+
+#endif
+}
+
+TEST_CASE("Interoperability between composite and component devices",
+          "[oneapi_composite_device]") {
+#ifndef SYCL_EXT_ONEAPI_COMPOSITE_DEVICE
+  SKIP(
+      "The sycl_ext_oneapi_composite device extension is not supported by an "
+      "imlementation");
```
Comment (maarquitos14) [inline tests/extension/oneapi_composite_device/more_complex_test_cases.cpp:86]:
```suggestion
      "implementation");
```

File: tests/extension/oneapi_composite_device/descendent_device.cpp (hunk 2 above)
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/descendent_device.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/aspects.cpp (hunk 4 above)
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/aspects.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/aspects.cpp (hunk 3 above)
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/aspects.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/device_info_descriptors.cpp (hunk 5 above)
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/device_info_descriptors.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/device_info_descriptors.cpp (hunk 6 above)
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/device_info_descriptors.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/device_info_descriptors.cpp (hunk 7 above)
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/device_info_descriptors.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp (hunk 8 above)
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp (hunk 9 above)
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp (hunk 10 above)
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp (hunk 11 above)
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp (hunk 12 above)
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp (hunk 13 above)
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/more_complex_test_cases.cpp (hunk 18 above)
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/more_complex_test_cases.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/more_complex_test_cases.cpp (hunk 19 above)
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/more_complex_test_cases.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/device_info_descriptors.cpp (hunk 7 above)
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/device_info_descriptors.cpp]: e0fb65acf35bb15898984a98fe203bc489b868bc

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp (hunk 16 above)
Comment (maarquitos14) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:140]: I don't understand what you mean, sorry. The highlighted piece of code firstly gets all the devices of the platform, and subsequently it adds all the devices returned by `sycl::device::get_devices()` to the same vector. What I mean is that I think there will be duplicates here. If we do add all the devices from `sycl::device::get_devices()`, why do we need to start with `platform.get_devices()`? Like I said, I would expect duplicates here.

File: tests/extension/oneapi_composite_device/CMakeLists.txt [hunk 20]
```diff
@@ -0,0 +1,11 @@
+if(SYCL_CTS_ENABLE_EXT_ONEAPI_COMPOSITE_DEVICE_TESTS)
+  set(test_cases_list
+    enumerating_composite_devices.cpp
+    device_info_descriptors.cpp
+    aspects.cpp
+    descendent_device.cpp
+    more_complex_test_cases.cpp
```
Comment (bader) [inline tests/extension/oneapi_composite_device/CMakeLists.txt:7]:
```suggestion
    aspects.cpp
    descendent_device.cpp
    device_info_descriptors.cpp
    enumerating_composite_devices.cpp
    more_complex_test_cases.cpp
```

[the remark this answers is not in the document]
Comment (bader) [reply tests/extension/oneapi_composite_device/CMakeLists.txt]: Done.

**Comment by maarquitos14** [comment]
@bader Other than the current unresolved threads, this LGTM, but I'd like to have either some discussion or resolution for the unresolved threads before approving.

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp (hunk 16 above)
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:140]: Apology, I misread the question and mistakenly though that it was asked about a different test.

I agree that `platform.get_devices()` should contain everything that `device::get_devices()` returns, so we can probably simplify the code here

Comment (maarquitos14) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:140]: > I agree that platform.get_devices() should contain everything that device::get_devices() returns, so we can probably simplify the code here

I think it's actually the other way around, isn't it? The result of `device::get_devices()` should contain the result of `platform.get_devices()`.

Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp:140]: > I think it's actually the other way around, isn't it? The result of `device::get_devices()` should contain the result of `platform.get_devices()`.

Right :)

File: tests/extension/oneapi_composite_device/aspects.cpp (hunk 1 above)
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/aspects.cpp:45]: Added this test case into "enumerating composite devices" in bcebe1ad87778bca975673cc67eafaf1a1c23c26

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp (hunk 16 above)
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp]: Applied in f8b1ca288cbb9f405080482cf9b73cd3ed38b2eb

File: tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp (hunk 17 above)
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/enumerating_composite_devices.cpp]: Applied in f8b1ca288cbb9f405080482cf9b73cd3ed38b2eb

**Review by maarquitos14 — APPROVED** [review]
Just a few nits. Otherwise LGTM.

File: tests/extension/oneapi_composite_device/more_complex_test_cases.cpp [hunk 21]
```diff
@@ -0,0 +1,232 @@
...
+
+  component_queue
+      .submit([&](sycl::handler& cgh) {
+        sycl::accessor acc(bufC, cgh, sycl::read_write);
+
+        cgh.parallel_for(sycl::range{count}, [=](sycl::id<1> it) {
+          acc[it] += ptrA[it] + ptrB[it];
+        });
+      })
+      .wait();  // why this wait is required?
```
[the remark this answers is not in the document]
Comment (AlexeySachkov) [reply tests/extension/oneapi_composite_device/more_complex_test_cases.cpp:145]: Not yet, I plan to craft an isolated reproducer and investigate. We likely have a bug in our intel/llvm implementation that in cases like this we don't properly gather all dependencies when creating host accessor

**Review by keryell — COMMENTED** [review]
That looks nice.
Still have this `wait();  // why is it required?` question.

**Comment by AlexeySachkov** [comment]
> Still have this wait();  // why is it required? question.

I will update the PR with a better wording around that comment once I have this issue with dependency tracking reported. I don't think that it should block this PR, because this bug is very likely unrelated to the composite device extension, but the wording we have should definitely be improved before merging. That's still in my TODO-list

**Comment by keryell** [comment]
> > Still have this wait();  // why is it required? question.
> 
> I will update the PR with a better wording around that comment once I have this issue with dependency tracking reported. I don't think that it should block this PR, because this bug is very likely unrelated to the composite device extension, but the wording we have should definitely be improved before merging. That's still in my TODO-list

Otherwise you remove the `wait()` and merge the PR, as a compelling argument to have DPC++ implementation improved before passing next conformance test?

**Comment by AlexeySachkov** [comment]
> > > Still have this wait();  // why is it required? question.
> > 
> > 
> > I will update the PR with a better wording around that comment once I have this issue with dependency tracking reported. I don't think that it should block this PR, because this bug is very likely unrelated to the composite device extension, but the wording we have should definitely be improved before merging. That's still in my TODO-list
> 
> Otherwise you remove the `wait()` and merge the PR, as a compelling argument to have DPC++ implementation improved before passing next conformance test?

Yeah, that's  a better approach. My local testing shows that it is most likely directly related to the extension and not a general issue with dependencies and in that case it is especially better to highlight it.

Removed those `.wait()` calls in 9dcded43d54ced3d2ad1b39121237df01b4cbc1f and resolved merge conflicts, this should be ready to merge once CI passes.

**Review by keryell — APPROVED** [review]
Thanks!

**Pull request merged**